src/daft-logical-plan/src/optimization/rules/extract_window_function.rs

```diff
@@ -0,0 +1,163 @@
+    fn collect_window_functions(expr: &ExprRef, result: &mut Vec<(ExprRef, WindowSpec)>) {
+        match expr.as_ref() {
+            Expr::Function { func, .. } => {
+                // If this is a window function, extract its window spec
+                if let FunctionExpr::Window(window_func) = func {
+                    result.push((expr.clone(), window_func.window_spec.clone()));
+                }
+            }
+            // Recursively check children
+            _ => {
+                for child in expr.children() {
+                    Self::collect_window_functions(&child, result);
+                }
+            }
+        }
+    }
+
+    /// Helper function to replace window function expressions with column references to the Window output
+    fn replace_window_functions(
+        expr: &ExprRef,
+        window_col_mappings: &[(ExprRef, String)],
+    ) -> DaftResult<ExprRef> {
+        // Use transform pattern similar to replace_monotonic_id in DetectMonotonicId
+        let transformed = expr.clone().transform(|e| {
+            // First, check if this expression is a window function that needs to be replaced
+            for (window_expr, col_name) in window_col_mappings {
+                if Arc::ptr_eq(&e, window_expr) {
+                    // Replace with a column reference and mark as transformed
+                    return Ok(Transformed::yes(resolved_col(col_name.clone())));
+                }
+            }
+            // Not a window function to replace directly
+            Ok(Transformed::no(e))
+        })?;
+
+        Ok(transformed.data)
+    }
+}
+
+impl OptimizerRule for ExtractWindowFunction {
+    fn try_optimize(&self, plan: Arc<LogicalPlan>) -> DaftResult<Transformed<Arc<LogicalPlan>>> {
+        plan.transform_down(|node| {
+            match node.as_ref() {
+                LogicalPlan::Project(project) => {
+                    // Check if any expression contains window functions
+                    if Self::contains_window_function(project) {
+                        // Extract window functions and their specs
+                        let window_funcs = Self::extract_window_functions(&project.projection);
+
+                        if !window_funcs.is_empty() {
+                            let sample_window_spec = &window_funcs[0].1;
+
+                            // Extract the window function expressions
+                            let window_function_exprs = window_funcs
+                                .iter()
+                                .map(|(expr, _)| expr.clone())
+                                .collect::<Vec<ExprRef>>();
+
+                            // Create a Window operation with the window functions
+                            let window_plan = Arc::new(LogicalPlan::Window(
+                                Window::try_new(
+                                    project.input.clone(),
+                                    window_function_exprs.clone(),
```

Comment:
Instead of renaming in the `Window` constructor, we should just rename our window_function_exprs to their semantic IDs here so that the logic for naming is centralized to this rule